File: python/leetcode/univalued-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
  sought_value = None
  def check_unival(self, root: TreeNode) -> bool:
    if root is None:
      return True
    if root.val != self.sought_value:
      return False
    return self.check_unival(root.left) and self.check_unival(root.right)
  def isUnivalTree(self, root: TreeNode) -> bool:
    self.sought_value = root and root.val
    return self.check_unival(root)

File: java/leetcode/univalued-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean checkUnival(TreeNode root, int soughtValue) {
      if (root == null) return true;
      if (root.val != soughtValue) return false;
      return checkUnival(root.left, soughtValue) && checkUnival(root.right, soughtValue);
    }

    public boolean isUnivalTree(TreeNode root) {
        int soughtValue = root != null ? root.val : 0;
        return checkUnival(root, soughtValue);
    }
}
